File: packages/python/src/tesote_sdk/v1/accounts.py
"""v1 accounts resource. Read-only on `/v1/accounts`."""

from __future__ import annotations

from typing import Any, Dict, Optional

from ..models import Account, AccountList
from ..transport import Transport


class AccountsResource:
    """Read-only accounts on `/v1/accounts`."""

    _PREFIX = "/v1/accounts"

    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    def list(
        self,
        *,
        page: Optional[int] = None,
        per_page: Optional[int] = None,
        include: Optional[str] = None,
        sort: Optional[str] = None,
        cache_ttl: Optional[float] = None,
    ) -> AccountList:
        """GET /v1/accounts -- page-based pagination, ETag-cached for 60s upstream."""
        query: Dict[str, Any] = {
            "page": page,
            "per_page": per_page,
            "include": include,
            "sort": sort,
        }
        response = self._transport.request("GET", self._PREFIX, query=query, cache_ttl=cache_ttl)
        body: Dict[str, Any] = response.json if isinstance(response.json, dict) else {}
        return AccountList.from_dict(body)

    def get(self, account_id: str, *, cache_ttl: Optional[float] = None) -> Account:
        """GET /v1/accounts/{id}."""
        response = self._transport.request(
            "GET", f"{self._PREFIX}/{account_id}", cache_ttl=cache_ttl
        )
        body: Dict[str, Any] = response.json if isinstance(response.json, dict) else {}
        return Account.from_dict(body)


__all__ = ["AccountsResource"]

File: packages/ruby/lib/tesote_sdk/v1/accounts.rb
module TesoteSdk
  module V1
    class Accounts
      INDEX_CACHE_TTL = 60      # 1 minute ETag (spec)
      SHOW_CACHE_TTL = 300      # 5 minutes ETag (spec)

      def initialize(transport)
        @transport = transport
      end

      # GET /v1/accounts
      def list(query = {}, opts: {})
        body = @transport.request('GET', 'accounts', query: query, opts: with_cache(opts, INDEX_CACHE_TTL))
        Models::AccountList.from_hash(body)
      end

      # GET /v1/accounts/{id}
      def get(id, opts: {})
        raise ArgumentError, 'id is required' if id.nil? || id.to_s.empty?

        body = @transport.request('GET', "accounts/#{id}", opts: with_cache(opts, SHOW_CACHE_TTL))
        Models::Account.from_hash(body)
      end

      # GET /v1/accounts/{id}/transactions
      def list_transactions(account_id, query = {}, opts: {})
        raise ArgumentError, 'account_id is required' if account_id.nil? || account_id.to_s.empty?

        body = @transport.request('GET', "accounts/#{account_id}/transactions", query: query, opts: opts)
        Models::TransactionList.from_hash(body)
      end

      # Enumerate all transactions across pages (cursor-based).
      # Yields TransactionList page objects.
      def each_transaction_page(account_id, query = {}, opts: {}, &block)
        return enum_for(:each_transaction_page, account_id, query, opts: opts) unless block

        enum = Pagination::CursorEnumerator.new(start_query: query) do |q|
          list_transactions(account_id, q, opts: opts)
        end
        enum.each(&block)
      end

      private

      def with_cache(opts, ttl)
        return opts if opts.key?(:cache)

        opts.merge(cache: { ttl: ttl })
      end
    end
  end
end
